### ISSUE TITLE
Heartbeat is currently run in main thread, can it be moved to its own green thread context maybe using setTimeout ?

### ISSUE THREAD

**Issue by satishck1992** [tulios/kafkajs#1052]
**Is your feature request related to a problem? Please describe.**
A clear and concise description of what the problem is.
Right now if we are using eachMessage, and the consumer takes more time (maybe API endpoint got degraded etc), heartbeat is not sent, until the message is processed. This is not problematic in case of a single consumer, but in case of multiple consumers the amount of rebalancing which happens is huge. All the consumers of the group have to rejoin (until the time static membership has not been incorporated, and the group is using static membership) . In a way this solves issues like https://github.com/tulios/kafkajs/issues/387. This is also implemented in https://cwiki.apache.org/confluence/display/KAFKA/KIP-62%3A+Allow+consumer+to+send+heartbeats+from+a+background+thread.

In a way I kind of agree with the warning by @JaapRood , but differ in the conclusion. While the heartbeat is the status signal for the main consumer thread, it is mixing the consumption of the message with the health signal of the thread.  

**Describe the solution you'd like**
Instead of calling heartBeat in the main eachMessage loop, or exposing in eachBatch, we can write an internal timer which can beat at given intervals. 

A clear and concise description of what you want to happen. Please include links to other clients' solutions and/or KIPs for Kafka features if applicable.

https://cwiki.apache.org/confluence/display/KAFKA/KIP-62%3A+Allow+consumer+to+send+heartbeats+from+a+background+thread

**Additional context**

**Comment by JaapRood** [comment]
FYI, my [warning](https://github.com/tulios/kafkajs/pull/473#issuecomment-566947785) about how this is counter productive:

>> Can't we just send heartbeat outside of the loop by setTimeout or setInterval?

> You totally could, and the broker would accept it, but as soon as you call heartbeat outside of the processing loop, it stops being a good indicator of health of that loop. The heartbeat is at the heart (🙌) of Kafka's failure detection mechanism, which in turn is vital for the fault tolerance behaviour it tries to implement.

---

> While the heartbeat is the status signal for the main consumer thread, it is mixing the consumption of the message with the health signal of the thread.

Yes, it's mixing it, that's why it is effective. The cited KPI actually calls out how a separate heartbeat API _would_ be more effective, and their choice not to use it is mostly vestigial.

**Comment by t-d-d** [comment]
The proposal describes how the Java client works, although it does have a max.poll.interval setting as well.  See [here](https://stackoverflow.com/a/39759329) for an explanation.  And it does make sense in single threaded/ async concurrency model as well.

I think the suggestion should be considered.  It would be less API and easier to consume (e.g. the example @satishck1992 gives.)

**Comment by tulios** [comment]
We talked about exposing a `heartbeat` method in the consumer, so it would be simple to implement your own heartbeat timer. We haven't nailed the specifics on how to interrupt the loop and propagate the errors; that's why we dropped it at the time.

**Comment by t-d-d** [comment]
Yes, this could be a way forward.  I think a consumer.heartbeat() call would never throw.  Instead it should set an error or rebalanceRequired state on the consumer group.  The fetch loop (and batch processing) could then check for and take action one those signals.

**Comment by EricMCornelius** [comment]
Exposing an external heartbeat method would be very useful for certain specific use cases as well.

Example:

I've got a task that needs to retrieve large batches of data for efficient compression and archiving of a data stream.

This implies a long maxWait and large minBytes setting.

However, because heartbeating is currently only available during the eachBatch processing loop, I'm relegated to setting a huge sessionTimeout interval  - necessarily longer than my maxWait in the event of low event rates.

This of course means that dead consumers from non-clean shutdowns then don't get trimmed reasonably quickly.
